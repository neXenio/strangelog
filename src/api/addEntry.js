// @flow

import { outputFileSync } from 'fs-extra';
import jsYaml from 'js-yaml';

import type { ConfigType, EntryType } from '../types';

export default function addEntry(
  { path, components }: ConfigType,
  entry: EntryType
): void {
  if (entry.component && !Object.keys(components).includes(entry.component)) {
    throw new Error(`Unknown component "${entry.component || ''}"`);
  }

  const date = new Date();
  const fsFriendlyDateTimeString = toFSFriendlyDateTime(date);
  const readableComponent = entry.component || 'all';
  const fileName = `${fsFriendlyDateTimeString}_${entry.kind}_${readableComponent}.yml`;

  outputFileSync(
    `${path}/next/${fileName}`,
    jsYaml.safeDump({
      dateTime: date.toISOString(),
      ...entry
    })
  );
}

function toFSFriendlyDateTime(date: Date): string {
  return date.toISOString().replace(/:/g, '-');
}