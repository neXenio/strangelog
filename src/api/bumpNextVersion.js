// @flow

import {
  join as joinPath,
  resolve as resolvePath
} from 'path';

import { sync as syncGlob } from 'glob';
import { moveSync } from 'fs-extra';

import type { ConfigType } from '../types';

import { multiToSingleLineString } from './utils';

export default function bumpNextVersion(
  { path }: ConfigType,
  nextVersionString: string
): void {
  ensureNextVersionHasEntries(path, nextVersionString);
  ensureVersionDoesNotExist(path, nextVersionString);

  moveSync(
    joinPath(path, 'next'),
    joinPath(path, nextVersionString)
  );
}

function ensureNextVersionHasEntries(
  changelogPath: string,
  nextVersionString: string
) {
  if (syncGlob(joinPath(changelogPath, 'next', '**/*')).length === 0) {
    throw new Error(multiToSingleLineString(`
      Cannot release version "next" as ${nextVersionString}
      because it does not contain any entries yet
      (${resolvePath(changelogPath, 'next')})`));
  }
}

function ensureVersionDoesNotExist(
  changelogPath: string,
  nextVersionString: string
) {
  if (syncGlob(joinPath(changelogPath, nextVersionString)).length > 0) {
    throw new Error(multiToSingleLineString(`
      Cannot release version "next" as ${nextVersionString}
      because that version already exists
      (${resolvePath(changelogPath, nextVersionString)})`));
  }
}