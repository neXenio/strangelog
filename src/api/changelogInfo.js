// @flow

import { join as joinPath } from 'path';

import { outputFileSync, existsSync, readFileSync } from 'fs-extra';
import jsYaml from 'js-yaml';

import type { ChangelogInfoType, ConfigType } from '../types';

export function getChangelogInfo(config: ConfigType): ChangelogInfoType {
  const infoFilePath = joinPath(config.path, 'info.yml');

  return existsSync(infoFilePath)
    ? jsYaml.safeLoad(readFileSync(infoFilePath))
    : { version: -1 };
}

export function saveChangelogInfo(
  config: ConfigType,
  newChangelogInfo: ChangelogInfoType
): void {
  const infoFilePath = joinPath(config.path, 'info.yml');

  return outputFileSync(
    infoFilePath,
    jsYaml.safeDump(newChangelogInfo)
  );
}
