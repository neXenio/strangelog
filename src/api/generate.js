// @flow

import type { ConfigType, ChangelogType } from '../types';
import defaultTemplate from '../templates/defaultTemplate';

import { stringifyVersion } from './utils';

function readableComponent(componentID: string, { components }: ConfigType): string {
  if (componentID === null) {
    return 'All';
  }

  return components[componentID];
}

export default function generate(
  config: ConfigType,
  changelog: ChangelogType
): string {
  return defaultTemplate({
    readableComponent: (componentID) => readableComponent(componentID, config),
    stringifyVersion
  }, changelog);
}
