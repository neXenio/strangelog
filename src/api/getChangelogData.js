// @flow

import { join as joinPath } from 'path';

import { readFileSync } from 'fs-extra';
import groupBy from 'lodash/groupBy';
import jsYaml from 'js-yaml';
import { sync as syncGlob } from 'glob';

import type {
  ConfigType,
  ChangelogType,
  VersionChangelogType,
  VersionType,
  EntryType
} from '../types';

import { stringifyVersion } from './utils';
import getSortedChangelogVersions from './getSortedChangelogVersions';

const entryKinds = [
  'addition',
  'change',
  'fix',
  'removal',
  'deprecation',
  'security',
];

export default function getChangelogData(
  config: ConfigType
): ChangelogType {
  return [
    getVersionChangelog(config, null),
    ...getSortedChangelogVersions(config).map((version) =>
      getVersionChangelog(config, version))
  ];
}

function getVersionChangelog(
  config: ConfigType,
  version: ?VersionType
): VersionChangelogType {
  return {
    version,
    entries: {
      ...entryKinds.reduce((kinds, entryKind) => ({
        ...kinds,
        [entryKind]: []
      }), {}),
      ...groupBy((
        getVersionChangelogFileNames(config, stringifyVersion(version))
          .map((entryFileName) => readFileSync(entryFileName).toString())
          .map(jsYaml.safeLoad)
          .sort(sortByComponent.bind(null, config))
      ), 'kind')
    }
  };
}

function sortByComponent(config: ConfigType, entry1: EntryType, entry2: EntryType) {
  if (!entry1.component && !entry2.component) {
    return 0;
  }

  if (!entry1.component && entry2.component) {
    return 1;
  }

  if (entry1.component && !entry2.component) {
    return -1;
  }

  // $FlowFixMe: Checks above cover that
  return config.components[entry1.component].localeCompare(config.components[entry2.component]);
}

function getVersionChangelogFileNames(
  { path }: ConfigType,
  versionString: string
): string[] {
  return syncGlob(joinPath(path, versionString, '*.yml'));
}