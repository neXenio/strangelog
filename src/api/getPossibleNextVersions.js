// @flow

import { readFileSync, existsSync } from 'fs-extra';
import { parse as parseSemVer } from 'semver';

import type { VersionType } from '../types';

export default function getPossibleNextVersions(): VersionType[] | null {
  const packageVersion = existsSync('package.json')
    ? parseSemVer(JSON.parse(readFileSync('package.json').toString()).version)
    : null;

  if (!packageVersion) {
    return null;
  }

  return [{
    major: packageVersion.major,
    minor: packageVersion.minor,
    patch: packageVersion.patch + 1
  }, {
    major: packageVersion.major,
    minor: packageVersion.minor + 1,
    patch: 0
  }, {
    major: packageVersion.major + 1,
    minor: 0,
    patch: 0
  }];
}
