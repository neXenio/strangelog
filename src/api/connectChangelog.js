// @flow

import type { ConfigType, ChangelogAPIType } from '../types';

import addEntry from './addEntry';
import bumpNextVersion from './bumpNextVersion';
import generate from './generate';
import getChangelogData from './getChangelogData';
import getPossibleNextVersions from './getPossibleNextVersions';
import migrate from './migrate';
import { getChangelogInfo, saveChangelogInfo } from './changelogInfo';

export default function connectChangelog(config: ConfigType): ChangelogAPIType {
  return {
    addEntry(entry) {
      return addEntry(config, entry);
    },
    bumpNextVersion(nextVersion) {
      return bumpNextVersion(config, nextVersion);
    },
    getChangelogData() {
      return getChangelogData(config);
    },
    generate() {
      return generate(config, getChangelogData(config));
    },
    getPossibleNextVersions() {
      return getPossibleNextVersions(config);
    },
    getComponentsConfig() {
      return config.components;
    },
    migrate() {
      return migrate(config);
    },
    getChangelogInfo() {
      return getChangelogInfo(config);
    },
    saveChangelogInfo(newChangelogInfo) {
      return saveChangelogInfo(config, newChangelogInfo);
    }
  };
}
