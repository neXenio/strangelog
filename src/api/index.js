// @flow

import connectChangelog from './connectChangelog';
import { CURRENT_VERSION } from './migrations';

export { connectChangelog, CURRENT_VERSION };
