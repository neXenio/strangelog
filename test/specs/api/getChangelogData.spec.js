// @flow

import { removeSync } from 'fs-extra';
import { install as installClock } from 'lolex';

import { connectChangelog } from '../../../src/api';
import { addTestVersionsWithEntries } from '../../factories/changelog';
import { getOwnTestPath } from '../../factories/fileSystem';

const testPath = getOwnTestPath();

describe('getChangelogData', () => {

  let clock;

  beforeEach(() => {
    removeSync(testPath);
    clock = installClock(new Date('2017-06-24T00:01:02.000Z'));
  });

  afterEach(() => {
    removeSync(testPath);
    clock.uninstall();
  });

  function setup() {
    return connectChangelog({
      path: testPath,
      components: {
        comp1: 'Comp 1',
        comp2: 'Comp 2',
        comp3: 'Comp 3'
      }
    });
  }

  it('returns correct changelog data', () => {
    const changelogAPI = setup();

    addTestVersionsWithEntries(changelogAPI);

    expect(changelogAPI.getChangelogData()).toMatchSnapshot();
  });

  it('sorts the entries by their component name', () => {
    const changelogAPI = setup();

    changelogAPI.addEntry({
      component: 'comp2',
      kind: 'addition',
      description: 'some silly description'
    });
    changelogAPI.addEntry({
      component: 'comp3',
      kind: 'addition',
      description: 'some silly description'
    });
    changelogAPI.addEntry({
      component: 'comp1',
      kind: 'addition',
      description: 'some silly description'
    });

    const additionEntries = changelogAPI.getChangelogData()[0].entries.addition;

    expect(additionEntries[0].component).toBe('comp1');
    expect(additionEntries[1].component).toBe('comp2');
    expect(additionEntries[2].component).toBe('comp3');
  });

  describe('when there are no entries of a certain kind', () => {

    it('returns an empty array for that kind', () => {
      const changelogAPI = setup();

      changelogAPI.addEntry({
        component: 'comp1',
        kind: 'removal',
        description: 'some removal description'
      });

      expect(changelogAPI.getChangelogData()[0].entries.deprecation).toEqual([]);
    });

  });

});
