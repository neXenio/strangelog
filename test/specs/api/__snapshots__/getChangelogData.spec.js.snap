// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getChangelogData returns correct changelog data 1`] = `
Array [
  Object {
    "entries": Object {
      "addition": Array [],
      "change": Array [],
      "deprecation": Array [],
      "fix": Array [
        Object {
          "component": "comp2",
          "dateTime": "2017-06-24T00:01:02.000Z",
          "description": "comp2 fix description",
          "kind": "fix",
        },
      ],
      "removal": Array [],
      "security": Array [],
    },
    "version": null,
  },
  Object {
    "entries": Object {
      "addition": Array [],
      "change": Array [
        Object {
          "component": "comp1",
          "dateTime": "2017-06-24T00:01:02.000Z",
          "description": "comp1 change description",
          "kind": "change",
        },
      ],
      "deprecation": Array [],
      "fix": Array [],
      "removal": Array [],
      "security": Array [],
    },
    "version": Object {
      "major": 1,
      "minor": 1,
      "patch": 0,
    },
  },
  Object {
    "entries": Object {
      "addition": Array [
        Object {
          "component": "comp1",
          "dateTime": "2017-06-24T00:01:02.000Z",
          "description": "comp1 addition description",
          "kind": "addition",
        },
      ],
      "change": Array [],
      "deprecation": Array [],
      "fix": Array [],
      "removal": Array [],
      "security": Array [],
    },
    "version": Object {
      "major": 1,
      "minor": 0,
      "patch": 0,
    },
  },
]
`;
